manticore/core/cpu/binja.py

```diff
@@ -0,0 +1,1184 @@
+
+    def __init__(self, view, memory):
+        '''
+        Builds a CPU model.
+        :param arch: BinaryNinja arch.
+        '''
+        # get a platform specific CPU
+        platform_cpu = CpuFactory.get_cpu(memory, 'amd64')
+        # and mark it as virtual so as to not increment the PC in the
+        # @instruction decorator
+        platform_cpu.real_cpu = False
+        self.__class__.platform_cpu = platform_cpu
+        platform_regs = platform_cpu.all_registers
+        from binaryninja import Architecture
+        self.__class__.max_instr_width = view.arch.max_instr_length
+        self.__class__.address_bit_size = 8 * view.arch.address_size
+        self.__class__.machine = self.platform_cpu.machine
+        self.__class__.mode = self.platform_cpu.mode
+        self.__class__.arch = view.arch
+        # FIXME
+        self.__class__.disasm = BinjaILDisasm(view)
+        self.view = view
+        self._segments = {}
+
+        # initialize the memory and register files
+        # FIXME
+        super(BinjaCpu, self).__init__(BinjaRegisterFile('x86_64', platform_regs),
+                                       memory)
+
+    @property
+    def function_hooks(self):
+        return dict(self._function_hooks)
+
+    @property
+    def instr_hooks(self):
+        return defaultdict(list, self._instr_hooks)
+
+    def __getstate__(self):
+        state = super(BinjaCpu, self).__getstate__()
+        state['segments'] = self._segments
+        return state
+
+    def __setstate__(self, state):
+        self._segments = state['segments']
+        super(BinjaCpu, self).__setstate__(state)
+
+    def canonicalize_instruction_name(self, insn):
+        if isinstance(insn, BinjaILDisasm.BinjaILInstruction):
+            return insn.name
+        # fallback
+        return self.platform_cpu.canonicalize_instruction_name(insn)
+
+    def set_descriptor(self, selector, base, limit, perms):
+        assert selector > 0 and selector < 0xffff
+        assert base >= 0 and base < (1 << self.address_bit_size)
+        assert limit >= 0 and limit < 0xffff or limit & 0xfff == 0
+        self._segments[selector] = (base, limit, perms)
+
+    def get_descriptor(self, selector):
+        return self._segments.setdefault(selector, (0, 0xfffff000, 'rwx'))
+
+    def _wrap_operands(self, operands):
+        import binaryninja.enums as enums
+
+        # if we don't support this, fallback to default platform :( :(
+        # we will be using capstone anyhow, as we need to move on..
+        o = self.disasm.disasm_il.operation
+        if (o == enums.LowLevelILOperation.LLIL_UNIMPL or
+                o == enums.LowLevelILOperation.LLIL_UNIMPL_MEM):
+            return [AMD64Operand(self.platform_cpu, op) for op in operands]
+
+        return [BinjaOperand(self, self.disasm.disasm_il, op)
+                for op in operands]
+
+    # XXX this is currently not active because a bunch of flag-setting
+    # LLIL are not implemented by Binja :(
```

Comment:
we should also comment what actually is active also, so we have a better idea of where it *would* fit in when we revisit this